### PR TITLE
Fix MVC form data binding localization

### LINKED ISSUE #6566: Localised number format and wrong input type for decimal fields
### Describe the bug
With a model that contains a `decimal` property, this doesn't work with a locale that uses a decimal comma:

    <input asp-for="Number">
    <input asp-for="Number" type="number">

### To Reproduce
1. Using this version of ASP.NET Core 2.1.5
2. Create the model class and the HTML razor view, then run it

Code:

    class Model
    {
        public decimal Number { get; set; } = 1.0m;
    }

### Expected behavior
This should be the rendered HTML:

    <input type="number" name="Number" value="1.0">

This happens instead, for the two view variants from the top:

    <input type="text" name="Number" value="1,0">
    <input type="number" name="Number" value="1,0">

The first issue is that the `decimal` type isn't rendered with the `type="number"` attribute. If I fix this myself by adding the attribute, the field remains completely empty because there is a **comma** instead of a **point** in the value. This shouldn't be localised because then no browser or frontend library can handle the value anymore.

PS: I haven't even got to try what happens when the correct value "1.0" is sent back to the controller and the model binder should convert it to a `decimal` type. It probably fails for the same reason.

[User reference: Configuration/Endpoint/Edit/Factor]

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# Fix MVC form data binding localization

This PR fixes an issue in which all form elements were being formatted and parsed using the "current" culture, while some form elements require culture-invariant formatting and parsing.

## Description

Some form elements (e.g., `<input type="text"/>`) require culture-specific formatting and parsing because their values are directly entered by the user. However, other inputs (e.g., `<input type="number"/>`) use culture-invariant formatting both in the HTML source and in the form request. Thus, we need to ensure that the correct formatting will be applied for each type of form element.

The fix for this bug comes in two parts:
1. When generating HTML form elements, we use culture-invariant formatting for the `value` attribute when appropriate.
2. When parsing values in a form request, use the invariant culture when the value's corresponding form element uses culture-invariant formatting.

To achieve step 2, we generate an additional HTML element for all form elements that use invariant formatting:
`<input name="__Invariant" type="hidden" value="<model property name>"/>`

This way, the form request gives us the information we need about which inputs need culture-invariant parsing.

Fixes #6566

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/MvcOptions.cs [hunk 1]
```diff
@@ -48,6 +50,13 @@ public MvcOptions()
         ModelMetadataDetailsProviders = new List<IMetadataDetailsProvider>();
         ModelValidatorProviders = new List<IModelValidatorProvider>();
         ValueProviderFactories = new List<IValueProviderFactory>();
+
+        _allowCultureInvariantFormModelBinding = new(nameof(AllowCultureInvariantFormModelBinding));
+
+        _switches = new ICompatibilitySwitch[]
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.Core/src/MvcOptions.cs:56]: Note: the mechanism around opting in/out of this fix is still a WIP.

File: src/Mvc/Mvc.TagHelpers/src/InputTagHelper.cs [hunk 2]
```diff
@@ -410,6 +431,16 @@ private TagBuilder GenerateRadio(ModelExplorer modelExplorer, IDictionary<string
             htmlAttributes);
     }
 
+    private static TagBuilder GenerateInvariantCultureMetadata(string propertyName)
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.TagHelpers/src/InputTagHelper.cs:434]: It would be nice to use `IHtmlGenerator` for this, but introducing new methods there would be a breaking change.

File: src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs [hunk 3]
```diff
@@ -51,9 +52,11 @@ public class DefaultHtmlGenerator : IHtmlGenerator
     /// <param name="urlHelperFactory">The <see cref="IUrlHelperFactory"/>.</param>
     /// <param name="htmlEncoder">The <see cref="HtmlEncoder"/>.</param>
     /// <param name="validationAttributeProvider">The <see cref="ValidationHtmlAttributeProvider"/>.</param>
+    /// <param name="mvcOptionsAccessor">The accessor for <see cref="MvcOptions"/>.</param>
     public DefaultHtmlGenerator(
         IAntiforgery antiforgery,
         IOptions<MvcViewOptions> optionsAccessor,
+        IOptions<MvcOptions> mvcOptionsAccessor,
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs:59]: I realize this is a breaking change. Giving this type access to the `MvcOptions` in an alternate manner is something I still need to investigate.

File: src/Mvc/Mvc.ViewFeatures/src/ViewDataDictionary.cs [hunk 4]
```diff
@@ -391,18 +391,30 @@ public ICollection<string> Keys
...
+    /// <summary>
+    /// Formats the given <paramref name="value"/> using the given <paramref name="format"/> and <paramref name="formatProvider"/>.
+    /// </summary>
+    /// <param name="value">The value to format.</param>
+    /// <param name="format">
+    /// The format string (see <see href="https://msdn.microsoft.com/en-us/library/txafckwd.aspx"/>).
+    /// </param>
+    /// <param name="formatProvider">The <see cref="IFormatProvider"/> used to format the value.</param>
+    /// <returns>The formatted <see cref="string"/>.</returns>
+    public static string? FormatValue(object? value, string? format, IFormatProvider formatProvider)
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.ViewFeatures/src/ViewDataDictionary.cs:417]: This new method (as well as some of the other public API additions) could possibly be made `internal`. We need to decide which of these new APIs (if any) should be public.

File: src/Shared/ModelBinding/FormModelBindingHelper.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+namespace Microsoft.AspNetCore.Mvc.ModelBinding;
+
+internal static class FormModelBindingHelper
+{
+    public const string CultureInvariantFieldName = "__Invariant";
+
+    public static bool InputTypeUsesCultureInvariantFormatting(string? inputType)
+        => inputType switch
+        {
+            "number" => true,
```
Comment (MackinnonBuck) [inline src/Shared/ModelBinding/FormModelBindingHelper.cs:15]: There are other inputs requiring invariant formatting, still need to add them here. In fact, I'm realizing as I'm writing this that it's probably the _majority_ of `<input />` types that would use invariant formatting. If this is the case, we may want to invert this and instead emit an extra form field for each occurrence of a culture-specific entry.

**Review by MackinnonBuck — COMMENTED** [review]
Just adding a few comments to give some context in case there are any early reviewers.

File: src/Mvc/Mvc.TagHelpers/src/InputTagHelper.cs (hunk 2 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.TagHelpers/src/InputTagHelper.cs:420]: However, we _could_ add it as an optional interface member if we're ok with adding more to the public API surface.

File: src/Shared/ModelBinding/FormModelBindingHelper.cs (hunk 5 above)
Comment (MackinnonBuck) [reply src/Shared/ModelBinding/FormModelBindingHelper.cs]: Update: I've just added all of the input types that use invariant formatting. Maybe we could scrutinize this list and eliminate input types where either culture-specific or culture-invariant parsing/formatting works (like in the checkbox case). That makes this change affect as few inputs as possible. We would just need to make sure we're not put in a situation where we need to modify this list later, since that would be another breaking change.

**Comment by MackinnonBuck** [comment]
Note that the way this is currently implemented, the new behavior is the default for all existing and new projects. In order to opt-out and revert to the old behavior, apps will need to do something like this:

```csharp
var builder = services.AddMvc()
    // ...
    .AddMvcOptions(options =>
    {
        options.SuppressCultureInvariantFormModelBinding = true;
    })
    .AddViewOptions(options =>
    {
        options.HtmlHelperOptions.SuppressCultureInvariantFormValueFormatting = true;
    });
```

There are two options to configure because one controls the "model" end of things (parsing form values) and one controls the HTML generation (formatting form element values). We could attempt to unify these, but this would probably require breaking changes to our public API (passing an `MvcOptions`) object through to the tag helper/HTML generation classes.

The decision to make this fix "opt-out" is not final. I'd love to hear more opinions about what the right choice is.

**Comment by brunolins16** [comment]
> Note that the way this is currently implemented, the new behavior is the default for all existing and new projects. In order to opt-out and revert to the old behavior, apps will need to do something like this:
> 
> ```cs
> var builder = services.AddMvc()
>     // ...
>     .AddMvcOptions(options =>
>     {
>         options.SuppressCultureInvariantFormModelBinding = true;
>     })
>     .AddViewOptions(options =>
>     {
>         options.HtmlHelperOptions.SuppressCultureInvariantFormValueFormatting = true;
>     });
> ```
> 
> There are two options to configure because one controls the "model" end of things (parsing form values) and one controls the HTML generation (formatting form element values). We could attempt to unify these, but this would probably require breaking changes to our public API (passing an `MvcOptions`) object through to the tag helper/HTML generation classes.
> 
> The decision to make this fix "opt-out" is not final. I'd love to hear more opinions about what the right choice is.

How bad it is if only one of them is set?

**Comment by dougbu** [comment]
> How bad it is if the user sets only one of them?

Suspect the result of setting _only_ `SuppressCultureInvariantFormValueFormatting` would be browser errors or incorrect display (though the current culture would need to be problematic, and the value would need to be large or small enough to require punctuation. In the `SuppressCultureInvariantFormModelBinding` case, again with problematic cultures and values, round trips would fail.

**Comment by dougbu** [comment]
> How bad it is if the user sets only one of them?

This is worth testing @MackinnonBuck. The follow-on question is basically, "Should we instead have one `MvcOptions.SuppressCultureInvariantRoundTripping` option❔"

File: src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs [hunk 6]
```diff
@@ -1329,27 +1339,15 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
 
             case InputType.Text:
             default:
-                var attributeValue = (string)GetModelStateValue(viewContext, fullName, typeof(string));
-                if (attributeValue == null)
-                {
-                    attributeValue = useViewData ? EvalString(viewContext, expression, format) : valueParameter;
-                }
```
Comment (dougbu) [inline src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs:1336]: Moving this around feels like an unrelated optimization. Am I missing something❔

File: src/Mvc/Mvc.Core/src/ModelBinding/FormModelBindingHelper.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding;
+
+internal static class FormModelBindingHelper
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/ModelBinding/FormModelBindingHelper.cs:6]: Suggest choosing a more general name since this is also used in HTML generation

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs [hunk 8]
```diff
@@ -39,9 +42,30 @@ public class FormValueProvider : BindingSourceValueProvider, IEnumerableValuePro
...
         Culture = culture;
     }
 
+    /// <summary>
+    /// Creates a value provider for <see cref="IFormCollection"/>.
+    /// </summary>
+    /// <param name="bindingSource">The <see cref="BindingSource"/> for the data.</param>
+    /// <param name="values">The key value pairs to wrap.</param>
+    /// <param name="culture">The culture to return with ValueProviderResult instances.</param>
+    public FormValueProvider(
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:61]: nit; We usually put the constructors w/ more arguments later in the `class`.

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProviderFactory.cs [hunk 9]
```diff
@@ -14,6 +14,25 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding;
...
+    /// <param name="options">The <see cref="MvcOptions"/> options.</param>
+    public FormValueProviderFactory(MvcOptions? options)
+    {
+        _options = options;
+    }
+
+    /// <summary>
+    /// Creates a new <see cref="FormValueProviderFactory"/>.
+    /// </summary>
+    public FormValueProviderFactory()
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/ModelBinding/FormValueProviderFactory.cs:31]: Same nit: Suggest moving this constructor above the other one.

File: src/Mvc/Mvc.Core/src/MvcOptions.cs [hunk 10]
```diff
@@ -379,6 +380,19 @@ public int MaxModelBindingRecursionDepth
     /// <value>Defaults to <c>8192</c>.</value>
     public int MaxIAsyncEnumerableBufferLimit { get; set; } = 8192;
 
+    /// <summary>
+    /// Gets or sets a value that determines if form values are disallowed to be
+    /// parsed using <see cref="CultureInfo.InvariantCulture"/>.
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/MvcOptions.cs:385]: I'm not sure how to word this but the current culture may be invariant, meaning "disallowed" is not correct. The general sense is more like "ignore `__InvariantCulture` and do not force associated fields to be parsed using `InvariantCulture`".

**Review by dougbu — COMMENTED** [review]
Kinda just starting&hellip;

**Comment by MackinnonBuck** [comment]
> How bad it is if the user sets only one of them?

If you only set `SuppressCultureInvariantFormModelBinding`, then the behavior remains the same as it is today from the perspective of the server, since the extra `__Invariant` form entries will be ignored in model binding. The main difference is that the generated HTML will not include incorrectly-formatted `value` attributes.

Only setting `SuppressCultureInvariantFormValueFormatting` means the both the HTML generation and model binding will behave exactly as they already to today (since model binding logic only changes when `__Invariant` entries are included in the form submission).

So, AFAICT, the behavior won't get _worse_ if you only set one of these options. Either you will get a slight improvement (generating better HTML), or no difference in behavior at all.

> Should we instead have one `MvcOptions.SuppressCultureInvariantRoundTripping` option❔

This is what I was hoping to do initially, but I couldn't find a good way to pass an `MvcOptions` through to the `DefaultHtmlGenerator` or `InputTagHelper`. It's not as simple as adding a new constructor (like I did for `FormValueProvider` and `FormValueProviderFactory`), because they're instantiated with DI and multiple constructors aren't permitted in that scenario.

File: src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs (hunk 6 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs]: Yes, it's unrelated. I originally changed this because an earlier prototype modified this part of the code, and the duplicated logic/state was getting in the way. Happy to change it back if there's any concern with the change.

File: src/Mvc/Mvc.Core/src/ModelBinding/FormModelBindingHelper.cs (hunk 7 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormModelBindingHelper.cs]: I've reorganized things a bit so the decision making around what should use "invariant" formatting doesn't occur in multiple places in the code (it now all exists in `DefaultHtmlGenerator`, and its decision is cached in `FormContext` so the `InputTagHelper` knows to generate the extra hidden element). As such, I've removed this type and placed the `CultureInvariantFieldName` string in `FormValueProvider`. Does that sound fine?

File: src/Mvc/Mvc.Core/src/MvcOptions.cs (hunk 10 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/MvcOptions.cs]: I've updated it to this:

> Gets or sets a value that determines if `<see cref="CultureInfo.CurrentCulture"/>` should always be used to parse form values.

Likewise, we could consider an alternate property name. Maybe `AlwaysUseCurrentCultureInFormModelBinding`?

**Comment by brunolins16** [comment]
> Only setting `SuppressCultureInvariantFormValueFormatting` means the both the HTML generation and model binding will behave exactly as they already do today (since model binding logic only changes when `__Invariant` entries are included in the form submission).
> 
> So, AFAICT, the behavior won't get _worse_ if you only set one of these options. Either you will get a slight improvement (generating better HTML), or no difference in behavior at all.

So, in this case, do we really need the `MvcOptions.SuppressCultureInvariantFormModelBinding`?

**Comment by MackinnonBuck** [comment]
> So, in this case, do we really need the `MvcOptions.SuppressCultureInvariantFormModelBinding`?

True, we might be able to go without it. I suppose the only use I can think of for it would be if you want the new behavior for generating correctly-formatted HTML `value` attribute strings, but not the new model binding behavior (in case someone was relying on the old behavior). Whether this is a real scenario, I'm not sure. And you could still set `SuppressCultureInvariantFormValueFormatting` to `true` as a workaround if you're willing to live with the existing `value` attribute formatting bug.

**Comment by brunolins16** [comment]
> EDIT: To cover the above scenario, we could possibly have one enum option, maybe something like HtmlHelperOptions.FormInputFormattingMode:

I don't know much about `HtmlHelperOptions` but this option is very interesting.

**Comment by MackinnonBuck** [comment]
`HtmlHelperOptions` is just where the existing options controlling HTML generation live. It's where the `SuppressCultureInvariantFormValueFormatting` property exists as the PR currently stands. So we would essentially be replacing that with an `enum` property and removing `SuppressCultureInvariantFormModelBinding`.

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs [hunk 11]
```diff
@@ -104,7 +130,8 @@ public override ValueProviderResult GetValue(string key)
         }
         else
         {
-            return new ValueProviderResult(values, Culture);
+            var culture = _invariantValueKeys?.Contains(key) == true ? CultureInfo.InvariantCulture : Culture;
```
Comment (brunolins16) [inline src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:133]: Can you add unit tests for this new behavior?

[the remark this answers is not in the document]
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:116]: To be honest, I don't have much knowledge/context about how JQuery fits into the picture here. I'm guessing if JQuery is being used, our tag helpers are not being used to generate the form? In this case, we would want to document the `__Invariant` string and make the same change to `JQueryFormValueProvider` so those using JQuery can take advantage of the new model binding logic. This is assuming JQuery doesn't do any of its own normalization logic to

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs [hunk 12]
```diff
@@ -143,4 +158,20 @@ public void RenderedField(string fieldName, bool value)
...
+    internal bool InvariantField(string fieldName)
+    {
+        ArgumentNullException.ThrowIfNull(fieldName);
+
+        InvariantFields.TryGetValue(fieldName, out var result);
+
+        return result;
+    }
+
+    internal void InvariantField(string fieldName, bool value)
```
Comment (brunolins16) [inline src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:171]: nit; It is internal but maybe you could call it `MarkAsInvariantField` or something similar to let clear what it is doing.

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs [hunk 13]
```diff
@@ -143,4 +158,20 @@ public void RenderedField(string fieldName, bool value)
 
         RenderedFields[fieldName] = value;
     }
+
+    internal bool InvariantField(string fieldName)
```
Comment (brunolins16) [inline src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:162]: nit: This one is clearer but maybe you could change to something like `IsInvariantField`?

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs [hunk 14]
```diff
@@ -107,6 +109,19 @@ public IList<IHtmlContent> EndOfFormContent
         }
     }
 
+    /// <summary>
+    /// Gets a dictionary mapping full HTML field names to indications that corresponding value was formatted
+    /// using <see cref="CultureInfo.InvariantCulture"/>.
+    /// </summary>
+    private Dictionary<string, bool> InvariantFields
```
[the remark this answers is not in the document]
Comment (MackinnonBuck) [reply src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:116]: This was for consistency with the `RenderedFields` property in this same class. I agree a hash set would probably suffice. Happy to change this if we're ok with the inconsistency.

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs (hunk 12 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:171]: I did it this way to be consistent with `RenderedField` in this same class. Again, happy to change if we're ok with using slightly inconsistent naming.

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs (hunk 13 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:162]: (as with the last two cases, this was for consistency)

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs (hunk 12 above)
Comment (brunolins16) [reply src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:171]: Oh! I should have checked the entire class before commenting 😂. Still unclear for me but I have never looked at this deeply before, so, I don't have a ton of context.

File: src/Mvc/Mvc.ViewFeatures/src/FormContext.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (brunolins16) [reply src/Mvc/Mvc.ViewFeatures/src/FormContext.cs:116]: Unfortunately, changing here, without change the methods, might not make sense. Probably is better keep the consistency.

**Review by dougbu — COMMENTED** [review]
Some of the comments may be out of date, sorry

File: src/Mvc/Mvc.Core/src/MvcOptions.cs [hunk 15]
```diff
@@ -379,6 +380,19 @@ public int MaxModelBindingRecursionDepth
     /// <value>Defaults to <c>8192</c>.</value>
     public int MaxIAsyncEnumerableBufferLimit { get; set; } = 8192;
 
+    /// <summary>
+    /// Gets or sets a value that determines if <see cref="CultureInfo.CurrentCulture"/> should always
+    /// be used to parse form values.
```
[the remark this answers is not in the document]
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/MvcOptions.cs]: If this property is `true`, then `CurrentCulture` will always be used to parse form values, regardless of the `type` attribute value. But, if we made this property the inverse of what it is currently (where a value of `true` implies the new behavior), then I think it would make sense to apply this suggestion (and also change `CurrentCulture` to `InvariantCulture`). There is some discussion in the [original issue](https://github.com/dotnet/aspnetcore/issues/6566#issuecomment-1213428561) about changing the name to `CultureInvariantFormValueFormattingEnabled`. Maybe we should do that and then apply this suggestion?

[the remark this answers is not in the document]
Comment (dougbu) [reply src/Mvc/Mvc.Core/src/MvcOptions.cs]: My nit was in error and I didn't remove it before submitting my Pending comments. Sorry.

I very much prefer `false` by default for `bool`s in our public API but am certainly willing to be ignored / overridden / corrected.

File: src/Mvc/Mvc.ViewFeatures/test/DefaultHtmlGeneratorTest.cs [hunk 16]
```diff
@@ -15,6 +15,7 @@
 using Microsoft.AspNetCore.Routing;
 using Microsoft.AspNetCore.Testing;
 using Microsoft.Extensions.Options;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
```
Comment (brunolins16) [inline src/Mvc/Mvc.ViewFeatures/test/DefaultHtmlGeneratorTest.cs:18]: Why we need this?

Comment (MackinnonBuck) [reply src/Mvc/Mvc.ViewFeatures/test/DefaultHtmlGeneratorTest.cs:18]: Added by mistake 🙃

File: src/Mvc/Mvc.Core/test/ModelBinding/FormValueProviderTest.cs [hunk 17]
```diff
@@ -9,6 +9,31 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding;
...
+        var valueProvider = GetEnumerableValueProvider(BindingSource.Query, values, culture);
+
+        // Act
+        var invariantCultureResult = valueProvider.GetValue(invariantCultureKey);
+        var currentCultureResult = valueProvider.GetValue(currentCultureKey);
+
+        // Assert
+        Assert.Equal(CultureInfo.InvariantCulture, invariantCultureResult.Culture);
+        Assert.Equal(BackingStore[invariantCultureKey], invariantCultureResult.Values);
+
```
Comment (brunolins16) [inline src/Mvc/Mvc.Core/test/ModelBinding/FormValueProviderTest.cs:32]: What do you think about split into 2 tests?

- `GetValue_ReturnsInvariantCulture_IfInvariantEntryExists`
- `GetValue_ReturnsCurrentCulture_IfIfInvariantEntryNotFound `-> you can find a better name 😁

Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/test/ModelBinding/FormValueProviderTest.cs:32]: We could do that, but the reason I had them included in the same test was so there are multiple form entries present, where some use culture-invariant formatting and others don't. Splitting it up into two tests would lose this detail.

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs [hunk 18]
```diff
@@ -13,7 +13,10 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding;
 /// </summary>
 public class FormValueProvider : BindingSourceValueProvider, IEnumerableValueProvider
 {
+    internal const string CultureInvariantFieldName = "__Invariant";
```
Comment (brunolins16) [inline src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: _**nit**_: I have no suggestions but would be great if it could be in another shared place instead of here. We are already using it in 3 different places

File: src/Mvc/Mvc.Core/test/ModelBinding/FormValueProviderTest.cs (hunk 17 above)
Comment (brunolins16) [reply src/Mvc/Mvc.Core/test/ModelBinding/FormValueProviderTest.cs:32]: Got it.

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs (hunk 18 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: Maybe in `HtmlHelperOptions`? We already have `_idAttributeDotReplacement` there.

Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: Actually, that won't work because `HtmlHelperOptions` is not accessible from `Mvc.Core`.

I wonder if we would want this string to be configurable, though? `IdAttributeDotReplacement` in `HtmlHelperOptions`, for example, can be overridden.

Comment (dougbu) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: Let's wait for someone to file an issue rather than proactively making things harder for ourselves.

File: src/Mvc/Mvc.TagHelpers/test/InputTagHelperTest.cs [hunk 19]
```diff
@@ -833,6 +833,75 @@ public void Process_GeneratesFormattedOutput_ForDateTime(string specifiedType, s
         Assert.Equal(expectedTagName, output.TagName);
     }
 
+    [Theory]
+    [InlineData("SomeProperty", "SomeProperty", true)]
+    [InlineData("SomeProperty", "[0].SomeProperty", true)]
+    [InlineData("SomeProperty", "[0].SomeProperty", false)]
```
Comment (dougbu) [inline src/Mvc/Mvc.TagHelpers/test/InputTagHelperTest.cs:839]: What about `SomePropertyA.SomePropertyB` e.g. src/Mvc/test/Mvc.FunctionalTests/compiler/resources/HtmlGenerationWebSite.HtmlGeneration_Home.Warehouse.html❔

File: src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs [hunk 20]
```diff
@@ -1329,27 +1347,15 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
 
             case InputType.Text:
             default:
-                var attributeValue = (string)GetModelStateValue(viewContext, fullName, typeof(string));
-                if (attributeValue == null)
+                if (string.Equals(suppliedTypeString, "file", StringComparison.OrdinalIgnoreCase) ||
+                    string.Equals(suppliedTypeString, "image", StringComparison.OrdinalIgnoreCase))
```
Comment (dougbu) [inline src/Mvc/Mvc.ViewFeatures/src/DefaultHtmlGenerator.cs:1351]: nit: suggest avoiding the empty block e.g.
```suggestion
                // 'value' attribute is not needed for 'file' and 'image' input types.
                if (!string.Equals(suppliedTypeString, "file", StringComparison.OrdinalIgnoreCase) &&
                    !string.Equals(suppliedTypeString, "image", StringComparison.OrdinalIgnoreCase))
                { ... }
```

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs (hunk 18 above)
Comment (MackinnonBuck) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: We determined in API review that it's not necessary to make the "__Invariant" string configurable.

There doesn't seem to be another good existing location to place this constant, so how about a new helper class, `FormValueHelper`?

**Review by dougbu — APPROVED** [review]
Nothing but nits 😄

File: src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs (hunk 18 above)
Comment (dougbu) [reply src/Mvc/Mvc.Core/src/ModelBinding/FormValueProvider.cs:16]: The string `__Invariant` exists here, once in src/Mvc/Mvc.TagHelpers/test/InputTagHelperTest.cs, and in generated HTML files. Am I missing something❔ If not, I suggest leaving things as-is.

**Comment by MackinnonBuck** [comment]
Merging now so this fix makes the cutoff - I'll open a follow-up PR that addresses the remaining nits.

**Comment by bassem-mf** [comment]
This change broke a lot of my CSS. The ':last-child' CSS selector no longer applies to the elements that have the "__Invariant" input added after them.
